gh-137197: Add SSLContext.set_ciphersuites to set TLS 1.3 ciphers (#137198)
python/cpython: Doc/library/ssl.rst

.. method:: SSLContext.set_ciphers(ciphers)

   Set the allowed ciphers for sockets created with this context when

>    connecting using TLS 1.2 and earlier.  It should be a string in the `OpenSSL

Reviewer: ```suggestion
   connecting using TLS 1.2 and earlier.  *ciphers* should be a string in the `OpenSSL
```

I think it's better to say what's "it" is. I personally understand but removing any possible ambiguity would be better anyway. Up to you though. Same for *ciphersuites* below

Author: Done.

   cipher list format <https://docs.openssl.org/master/man1/ciphers/>`_.

>    To set allowed TLS 1.3 ciphers, use :meth:`SSHContext.set_ciphersuites`.

Reviewer 2: ```suggestion
   To set allowed TLS 1.3 ciphers, use :meth:`SSLContext.set_ciphersuites`.
```

Author: Fixed - thanks for catching this.

>    below.  If no cipher can be selected (because compile-time options or other

Reviewer: ```suggestion
   To set allowed TLS 1.3 ciphers, use :meth:`SSLContext.set_ciphersuites`.
   If no cipher can be selected (because compile-time options or other
```

Author: Fixed.

   configuration forbids use of all the specified ciphers), an
   :class:`SSLError` will be raised.

   .. note::
      when connected, the :meth:`SSLSocket.cipher` method of SSL sockets will

>       return the negotiated cipher and associated TLS version.

Reviewer: The docstring actually says:

```py
        """Return the currently selected cipher as a 3-tuple ``(name,
        ssl_version, secret_bits)``."""
```

So it's not a pair but a triplet that is being returned (that's how it's implemented in C). I suggest that we make this doc change in a separate PR.

Author: Do you want to revert just the changes made for the existing set_cipher() "note" but leave in this "note" text for the new set_ciphersuites() function? Alternately, what do you think of changing both function descriptions to say something like "return information about the negotiated cipher" and leave the the docs for the SSLSocket.cipher() method to cover this, where it already describes the tuple?

Author: After some thought, I'm thinking of the following for the notes text:

   .. note::
      when connected, the :meth:`SSLSocket.cipher` method of SSL sockets will
      return details about the negotiated cipher.

Reviewer: That's a good alternative as we don't need to know the details of `ciphers` in `set_ciphers/set_ciphersuites`. Another PR will take care of fixing the description of `ciphers`.

Author: The current docs for ``ciphers()`` says:

     Returns a three-value tuple containing the name of the cipher being used, the
     version of the SSL protocol that defines its use, and the number of secret
     bits being used.  If no connection has been established, returns ``None``.

This looks like it could be ok.

Reviewer: Oh.. so yes it's already fine!

.. method:: SSLContext.set_ciphers(ciphers)

   Set the allowed ciphers for sockets created with this context when
   connecting using TLS 1.2 and earlier.  The *ciphers* argument should
   be a string in the `OpenSSL cipher list format
   <https://docs.openssl.org/master/man1/ciphers/>`_.

>    To set allowed TLS 1.3 ciphers, use :meth:`SSLContext.set_ciphersuites`.

Reviewer: Note to myself: I would have added a linebreak after this sentence to separate from "If no cipher..." but I think the entire page needs some edits, so let's not spend too much time on wording for now.

Author: Since this change is already changing the line right above that, I'll go ahead and add the blank line.

   If no cipher can be selected (because compile-time options or other
   configuration forbids use of all the specified ciphers), an
   :class:`SSLError` will be raised.

   .. note::
      When connected, the :meth:`SSLSocket.cipher` method of SSL sockets will
      return details about the negotiated cipher.

> .. method:: SSLContext.set_ciphersuites(ciphersuites)

Reviewer: This needs a `.. versionadded:: next` (sorry I forgot about it).

Author: Will do.

   Set the allowed ciphers for sockets created with this context when
   connecting using TLS 1.3.  The *ciphersuites* argument should be a
   colon-separate string of TLS 1.3 cipher names.  If no cipher can be
   selected (because compile-time options or other configuration forbids
   use of all the specified ciphers), an :class:`SSLError` will be raised.

   .. note::
      when connected, the :meth:`SSLSocket.cipher` method of SSL sockets will

>       return details about the negotiated cipher.

Reviewer: ```suggestion
   .. note::

      When connected, the :meth:`SSLSocket.cipher` method of SSL sockets will
      return details about the negotiated cipher.
```

You can also update the other note. I never saw that we had a missing capital letter.

Author: Fixed both instances.

The TLS 1.3 protocol behaves slightly differently than previous version
of TLS/SSL. Some new TLS 1.3 features are not yet available.

- TLS 1.3 uses a disjunct set of cipher suites.  All AES-GCM and ChaCha20

>   cipher suites are enabled by default.  To restrict which TLS1.3 ciphers

Reviewer: ```suggestion
  cipher suites are enabled by default.  To restrict which TLS 1.3 ciphers
```

Author: Thanks.

>   are allowed, the method :meth:`SSLContext.set_ciphersuites` should be

Reviewer: For posterity: I really hate that OpenSSL named it `SSL_CTX_set_ciphersuites` for TLS 1.3 and later, and `SSL_CTX_set_cipher_list` for TLS 1.2 and below. I *hope* that this won't be annoying to users.

Author: I was surprised what they did here at first and I'm not a big fan about the naming chosen, but I can see why they didn't want to try and extend the existing command. It allows all kinds of partial matches on the five individual elements that make up a cipher suite in TLS 1.2 and earlier, plus other special keywords like LOW/MEDIUM/HIGH. All of that is gone with TLS 1.3, and the new function is much simpler and only supports an exact match against the very small number of cipher suites defined in TLS 1.3.

I don't really know why they didn't include "_list" in the name of the OpenSSL function, given that this seems to be the convention for other such calls that take a string argument.

The good news here is that things like post-quantum crypto are going to drive migration to TLS 1.3, and at some point setting TLS 1.2 cipher suites will no longer matter. At that point, the previous `set_ciphers()` can be deprecated and eventually removed, with only `set_ciphersuites()` remaining.

  called instead of :meth:`SSLContext.set_ciphers`, which only affects
  ciphers in older TLS versions.  The method :meth:`SSLContext.get_ciphers`
  returns information about ciphers for both TLS 1.3 and earlier versions

>   and the method :meth:`SSLSocket.cipher` returns information about the

Reviewer: In this paragraph, I would say "The XXX method" instead of "The method XXX".

Author: Will do.

The TLS 1.3 protocol behaves slightly differently than previous version
of TLS/SSL. Some new TLS 1.3 features are not yet available.

- TLS 1.3 uses a disjunct set of cipher suites.  All AES-GCM and ChaCha20
  cipher suites are enabled by default.  To restrict which TLS 1.3 ciphers

>   are allowed, the method :meth:`SSLContext.set_ciphersuites` should be

Reviewer: ```suggestion
  are allowed, the :meth:`SSLContext.set_ciphersuites` method should be
```
